# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [ ]:
#2//8 = 0 , solo la parte entera

In [6]:
def NumeroBinario(numero):
    if type(numero) != int or numero < 0: return None
    elif numero == 0: return 0
    binario = ''
    div_resto=0
    while numero > 0:
        div_resto = numero % 2
        numero //= 2
        binario = str(div_resto) + binario
    return int(binario)

In [8]:
#el slice es para que de str
print(bin(8)[2:])

1000


In [7]:
NumeroBinario(8)

1000

In [16]:
# Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
# representación binaria. Debe recibir y devolver un valor de tipo entero.
# En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.

def NumeroBinario(numero):
    if not isinstance(numero, int) or numero < 0:  #si el numero no es int
        return None  # Si el número no es entero o es menor que cero, retorna None
    
    binario = ''  # Variable para almacenar el número binario resultante

    while numero > 0:  # Mientras el número sea mayor que cero
        residuo = numero % 2  # Calcula el residuo al dividir por 2
        binario = str(residuo) + binario  # Agrega el residuo a la izquierda del número binario
        numero //= 2  # Actualiza el número dividiéndolo por 2 (parte entera)

    if not binario:  # Si el número original era 0, su representación binaria es '0'
        return 0
    else:
        return int(binario)


In [18]:
NumeroBinario(25)

11001

In [23]:
#Ejercicio Nay
#Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
#representación binaria. Debe recibir y devolver un valor de tipo entero.
#En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.

def NumeroBinarioDec(numero):
    if not isinstance(numero, int) or numero < 0:
        return None
    
    if numero == 0:
        return 0
    
    binario = 0
    i = 0
    
    while numero > 0:
        residuo = numero % 2
        binario += residuo * (10 ** i)
        numero //= 2
        i += 1
    
    return binario


In [24]:
#Ejemplo:
NumeroBinario(25)

11001

2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [75]:
def decimal_binario_frac(fraccion):

    if fraccion <= 0 or fraccion >= 1:
        return None 

    resultado_binario ="." #Variable para almacenar el número binario resultante

    while fraccion > 0: # Mientras el número sea mayor que cero
        fraccion *= 2 #La fracción se multiplica *2 para obtener el valor que sigue
        if fraccion >= 1:
            resultado_binario= "1" + resultado_binario #Para actualizar el valor binario
            fraccion -=1 
        else:
            resultado_binario += "0"
        
        if len(resultado_binario) >= 32: #Establecer limite en los digitos binarios
            return "La representación binaria se repite o es muy larga para calcular"
    return resultado_binario
#Fracciones a convertir en binario
    fracciones= [1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9]
    for fracc in fracciones:
        resultado = decimal_binario_frac(fracc)
    print(f"La representación binaria de {fracc} es: {resultado}")
    

In [22]:
#ejercicio Nay
def decimal_binario_frac(fraccion):
    if fraccion <= 0 or fraccion >= 1:
        return None 

    resultado_binario = "0."  # Variable para almacenar el número binario resultante

    while fraccion > 0:  # Mientras el número sea mayor que cero
        fraccion *= 2  # La fracción se multiplica por 2 para obtener el valor siguiente
        if fraccion >= 1:
            resultado_binario += "1"  # Para actualizar el valor binario
            fraccion -= 1
        else:
            resultado_binario += "0"

        if len(resultado_binario) >= 32:  # Establecer límite en los dígitos binarios
            return "La representación binaria se repite o es muy larga para calcular"
    return resultado_binario

# Fracciones a convertir en binario
fracciones = [1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9]

for fraccion in fracciones:
    resultado = decimal_binario_frac(fraccion)
    print(f"Fraccion 1 / {1/fraccion} : {fraccion} y En binario: {resultado}")

Fraccion 1 / 2.0 : 0.5 y En binario: 0.1
Fraccion 1 / 3.0 : 0.3333333333333333 y En binario: La representación binaria se repite o es muy larga para calcular
Fraccion 1 / 4.0 : 0.25 y En binario: 0.01
Fraccion 1 / 5.0 : 0.2 y En binario: La representación binaria se repite o es muy larga para calcular
Fraccion 1 / 6.0 : 0.16666666666666666 y En binario: La representación binaria se repite o es muy larga para calcular
Fraccion 1 / 7.0 : 0.14285714285714285 y En binario: La representación binaria se repite o es muy larga para calcular
Fraccion 1 / 8.0 : 0.125 y En binario: 0.001
Fraccion 1 / 9.0 : 0.1111111111111111 y En binario: La representación binaria se repite o es muy larga para calcular


In [79]:
decimal_binario_frac(1/4)

'1.0'

Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111

In [21]:
#res profe 
def binarioFraccion(numero_decimal, limite_decimal=24): #(lala, por defecto por si no entra numero, tomará ese)
    if numero_decimal >= 1 or numero_decimal < 0: return None #ya no se requiere else porque con el return se acaba
    elif numero_decimal == 0: return 0
    lista_binarios = []
    while numero_decimal > 0 and limite_decimal >0:
        lista_binarios.append(int(numero_decimal * 2)) #es la parte entera
        numero_decimal = numero_decimal *2 - int(numero_decimal * 2) #multiplico por 2 y le resto la parte entera 
        limite_decimal -=1

    numero_binario = "0."
    for num in lista_binarios: #para cada numero en la lista 
        numero_binario += str(num) #concatenación de los str
    return numero_binario


In [ ]:
binarioFraccion(1/7,10)